### PR TITLE
Add ObjectStore::get_opts (#2241)

### LINKED ISSUE #2241: Get Request Precondition Support
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

There are 4 broadly supported request preconditions:

* If-Match
* If-Not-Match
* If-Not-Modified-Since
* If-Unmodified-Since

https://cloud.google.com/storage/docs/request-preconditions
https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-conditional-headers-for-blob-service-operations
https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#API_GetObject_RequestSyntax

**Describe the solution you'd like**

I would like to propose adding

```
#[derive(Debug, Default)]
struct GetOptions {
    pub if_match: Option<String>,
    pub if_not_match: Option<String>,
    pub if_not_modified_since: Option<DateTime<Utc>>,
    pub if_unmodified_since: Option<DateTime<Utc>>,
}

pub trait ObjectStore {
     async fn get_opts(&self, location: &Path, options: GetOptions) -> Result<GetResult>

     async fn get_range_opts(&self, location: &Path, range: Option<usize>, options: GetOptions) -> Result<Bytes>
     ...
}

pub enum Error {
    #[snafu(display("Precondition failed: {}", source))]
    Precondition {
        source: Box<dyn std::error::Error + Send + Sync + 'static>,
    }

    #[snafu(display("Not modified: {}", source))]
    NotModified {
        source: Box<dyn std::error::Error + Send + Sync + 'static>,
    }
}
```

**Describe alternatives you've considered**

We could not add support for this

**Additional context**

#2230

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2241

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/lib.rs [hunk 1]
```diff
@@ -558,6 +575,62 @@ pub struct ObjectMeta {
     pub e_tag: Option<String>,
 }
 
+/// Options for a get request
+#[derive(Debug, Default)]
+pub struct GetOptions {
+    /// Request will succeed if the ETag matches
```
Comment (alamb) [inline object_store/src/lib.rs:581]: Could you please define `ETag` somewhere? I am not familiar with the term and I think a summary would be nicer UX than a IETF RFC link)

Also could we summarize explicitly  what happens if the option is set bit the request doesn't match (e.g "request will _only_ succeed" (if that is indeed the case))

File: object_store/src/lib.rs [hunk 2]
```diff
@@ -346,11 +346,24 @@ pub trait ObjectStore: std::fmt::Display + Send + Sync + Debug + 'static {
...
     /// Return the bytes that are stored at the specified location.
-    async fn get(&self, location: &Path) -> Result<GetResult>;
+    async fn get(&self, location: &Path) -> Result<GetResult> {
+        self.get_opts(location, GetOptions::default()).await
+    }
+
+    /// Perform a get request with options
+    ///
+    /// Note: options.range will be ignored if [`GetResult::File`]
+    async fn get_opts(&self, location: &Path, options: GetOptions) -> Result<GetResult>;
```
Comment (alamb) [inline object_store/src/lib.rs:356]: I think this is a breaking API change and thus we will need to make it a major version number upgrade -- any existing `impl ObjectStore` will stop compiling. I see you have marked the API as such 👍

File: object_store/src/aws/mod.rs [hunk 3]
```diff
@@ -246,39 +247,26 @@ impl ObjectStore for AmazonS3 {
...
-                store: "S3",
+                store: STORE,
                 source: Box::new(source),
             })
             .boxed();
 
         Ok(GetResult::Stream(stream))
     }
 
-    async fn get_range(&self, location: &Path, range: Range<usize>) -> Result<Bytes> {
```
Comment (alamb) [inline object_store/src/aws/mod.rs:262]: Where did this go?

File: object_store/src/lib.rs [hunk 4]
```diff
@@ -558,6 +575,62 @@ pub struct ObjectMeta {
     pub e_tag: Option<String>,
 }
 
+/// Options for a get request
```
Comment (alamb) [inline object_store/src/lib.rs:578]: I think it would help if we mentioned "for example, request range"

**Review by alamb — APPROVED** [review]
Looks like a good idea to me -- thank you @tustvold 

I think some more documentation would help, but is not required

Also, what do you think about marking `ObjectStore::get` and `ObjectStore::get_range` as depreated (with a note to use `ObjectStore::get_opts`)? That would put us on a path towards a more minimal API.

**Comment by tustvold** [comment]
> what do you think about marking ObjectStore::get and ObjectStore::get_range as depreated

I think the fact we provide default implementations is probably sufficient, I would rather not force all users to use `GetOptions` as the API is a little more clunky

File: object_store/src/aws/mod.rs (hunk 3 above)
Comment (tustvold) [reply object_store/src/aws/mod.rs:262]: The default implementation is sufficient

**Comment by alamb** [comment]
> I think the fact we provide default implementations is probably sufficient, I would rather not force all users to use GetOptions as the API is a little more clunky

I think `get_range` might still be worth deprecating -- I agree that `get` is a nice convenience that we could/should keep

**Pull request merged**